rectools/metrics/dq.py

```diff
@@ -0,0 +1,315 @@
+    """
+
+    deep: bool = attr.ib(default=False)
+
+    def calc(self, reco: pd.DataFrame) -> float:
+        """
+        Calculate metric value.
+
+        Parameters
+        ----------
+        reco : pd.DataFrame
+            Recommendations table with columns `Columns.User`, `Columns.Item`, `Columns.Rank`.
+
+        Returns
+        -------
+        float
+            Value of metric (average between users).
+        """
+        per_user = self.calc_per_user(reco)
+        return per_user.mean()
+
+    def calc_per_user(self, reco: pd.DataFrame) -> pd.Series:
+        """
+        Calculate metric values for all users.
+
+        Parameters
+        ----------
+        reco : pd.DataFrame
+            Recommendations table with columns `Columns.User`, `Columns.Item`, `Columns.Rank`.
+
+        Returns
+        -------
+        pd.Series
+            Values of metric (index - user id, values - metric value for every user).
+        """
+        raise NotImplementedError()
+
+
+class RecoEmpty(_RecoDQMetric):
+    """
+    Presence of empty rows in recommendations table when `k` recommendations are required for each
+    user. This metric helps to identify situations when recommendation lists are not fully filled.
+    Specify `deep=False` to calculate share of user with any number of empty rows at first `k`
+    positions.
+    Specify `deep=True` to calculate mean share of empty rows for each user at first `k` positions.
+
+    Parameters
+    ----------
+    k : int
+        Number required recommendations  for each user that will be used to calculate metric.
+    deep: bool, default `False`
+        Whether to calculated detailed value of the metric for each user. Otherwise just the share of
+        users with identified empty rows will be returned (this is the default behaviour).
+
+    Examples
+    --------
+    >>> reco = pd.DataFrame(
+    ...     {
+    ...         Columns.User: [1, 1, 2, 2, 2, 3, 3, 3, 3, 3],
+    ...         Columns.Item: [1, 2, 1, 2, 3, 1, 2, 3, 4, 5],
+    ...         Columns.Rank: [1, 2, 1, 2, 3, 1, 2, 3, 4, 5],
+    ...     }
+    ... )
+    >>> RecoEmpty(k=1).calc_per_user(reco).values
+    array([0, 0, 0])
+    >>> RecoEmpty(k=4).calc_per_user(reco).values
+    array([1, 1, 0])
+    >>> RecoEmpty(k=4, deep=True).calc_per_user(reco).values
+    array([0.5 , 0.25, 0.  ])
+    """
+
+    def calc_per_user(self, reco: pd.DataFrame) -> pd.Series:
+        """
+        Calculate metric values for all users.
+
+        Parameters
+        ----------
+        reco : pd.DataFrame
+            Recommendations table with columns `Columns.User`, `Columns.Item`, `Columns.Rank`.
+
+        Returns
+        -------
+        pd.Series
+            Values of metric (index - user id, values - metric value for every user).
+        """
+        reco_k = reco.query(f"{Columns.Rank} <= @self.k").drop_duplicates(subset=[Columns.User, Columns.Rank])
+        all_users = reco[Columns.User].unique()
```

Comment:
Yes. We have "missing ranks" case in our tests. And we are supporting recommendations from external sources so we can't be sure what's in there. Since it's DQ - we need to make sure we are checking everything right.

Reply:
Discussed to be ok

Reply:
fixed